File: website/apps/sns-manager/src/app/clients/reddit.ts
import { SNSPost, SocialProvider, User } from '@kumi-arts/core';
import { Axios, AxiosResponse } from 'axios';
import { createClient } from './client';

export interface RedditPost extends SNSPost {
  subreddit: string;
}

export interface PrepareImageUpload {
  s3UploadLease: { action: string; fields: [{ name: string; value: string }] };
  asset: { asset_id: string };
}

export class RedditClient {
  private client: Axios;

  constructor() {
    this.client = createClient(SocialProvider.REDDIT, {
      baseURL: '/api/reddit',
    });
  }

  private async prepareImageUpload(
    subreddit: string,
    media: File
  ): Promise<PrepareImageUpload> {
    const body = `filepath=${media.name}&mimetype=${media.type}`;
    return this.handleResponse(
      await this.client.post(
        `/r/${subreddit}/api/widget_image_upload_s3`,
        body,
        {
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
          },
        }
      )
    );
  }

  private async uploadImage(
    { s3UploadLease }: PrepareImageUpload,
    media: File
  ): Promise<void> {
    const body = new FormData();
    s3UploadLease.fields.forEach((f) => {
      body.append(f.name, f.value);
    });
    body.append('file', media, media.name);

    return this.handleResponse(await this.client.post(`/custom/media`, body));
  }

  async postMedia({
    subreddit,
    media,
    title,
    text,
  }: RedditPost): Promise<string> {
    const mediaIds = [];

    if (media) {
      const prepare = await this.prepareImageUpload(subreddit, media);
      const x = await this.uploadImage(prepare, media);
      console.log(x);

      mediaIds.push(prepare.asset.asset_id);
    }

    // const body = new FormData();
    // body.append('title', media.title);
    // body.append('items');
    // body.append('sr', media.subreddit);
    // body.append('kind', 'image');

    // if (mediaIds.length > 1) {
    const x: any = this.handleResponse(
      await this.client.post(
        `/api/submit_gallery_post.json?api_type=json`,
        JSON.stringify({
          items: mediaIds.map((id) => ({ media_id: id })),
          kind: 'self',
          sr: subreddit,
          title: title,
          text: text,
        })
      )
    );

    return x.json?.data?.url || '';
    // }

    // const body = new URLSearchParams({
    //   kind: 'image',
    //   sr: subreddit,
    //   title: title,
    //   text: text,
    //   url: 'https://reddit-uploaded-media.s3-accelerate.amazonaws.com/rte_images/8ax4zcskwy891',
    // });

    // const x: any = this.handleResponse(
    //   await this.client.post(`/api/submit?api_type=json`, body.toString(), {
    //     headers: {
    //       'Content-Type': 'application/x-www-form-urlencoded',
    //     },
    //   })
    // );

    // return x.json?.data?.url || '';
  }

  private handleResponse<T>(res: AxiosResponse<T>): T {
    if (res.status !== 200) {
      throw new Error(JSON.stringify(res.data));
    }

    return res.data;
  }

  async getUser(): Promise<User> {
    const { data } = await this.client.get('/api/v1/me');
    if (!data.name) {
      throw new Error('Failed to get user');
    }

    return { id: data.name, name: data.name };
  }

  async querySubreddit(query: string): Promise<string[]> {
    const urlParam = new URLSearchParams({ query });

    const { data } = await this.client.get(
      `/api/subreddit_autocomplete_v2?${urlParam.toString()}`
    );

    return data.data.children
      .map(
        (child: Record<string, Record<string, string>>) =>
          child.data.display_name
      )
      .filter((x: string) => !!x);
  }
}

File: app/client/reddit.py
import requests as req
from flask_openapi3 import FileStorage

from app.model import RedditPost


class RedditClient:
    asset_url = 'https://reddit-uploaded-media.s3-accelerate.amazonaws.com'

    def __init__(self, token: str, user_agent='script:sns-manager:0.0.1 (by /u/illu11)'):
        self.baseURL = 'https://oauth.reddit.com'
        self.headers = {
            'Authorization': f'Bearer {token}',
            'User-Agent': user_agent,
        }

    def get(self, url):
        res = req.get(self.baseURL + url, headers=self.headers)
        res.raise_for_status()
        return res.json()

    def _return_data_if_no_error(self, res):
        res.raise_for_status()

        json = res.json()['json']
        errors = json['errors']

        if len(errors) > 0:
            print(errors)
            raise Exception(errors)

        return json['data']

    def _add_comment(self, post_id: str, text: str):
        body = {
            'text': text,
            'thing_id': post_id,
        }
        res = req.post(self.baseURL + '/api/comment?api_type=json',
                       body, headers=self.headers)
        data = self._return_data_if_no_error(res)
        print(f'Added comment to post {post_id}: {text}')

        return data['things'][0]['data']['id']

    def submit_post(self, post: RedditPost):
        print(f'----- Start submitting to {post.subreddit} ------')

        image_ids = [self._upload_image(i) for i in post.images]
        post_id = self._post_to_subreddit(image_ids, post)

        if post_id and len(image_ids) > 0:
            self._add_comment(post_id, post.text)

        print(f'----- Submit finished -----')
        return post_id

    def _get_mimetype(self, image: FileStorage) -> str:
        return image.mimetype if image.mimetype else f'image/{image.filename.split(".")[-1]}'

    def _upload_image(self, image: FileStorage):
        data = {
            'filepath': image.filename,
            'mimetype': self._get_mimetype(image),
        }

        print(image.mimetype)

        res = req.post(self.baseURL + '/api/media/asset.json',
                       data=data, headers=self.headers)
        res.raise_for_status()

        asset_data = res.json()
        self._upload_image_action_to_s3(image, asset_data['args'])

        asset_id = asset_data['asset']['asset_id']
        print('Uploaded {} with id {}'.format(image.filename, asset_id))
        return asset_id

    def _upload_image_action_to_s3(self, image: FileStorage, args: dict):
        full_url = 'https:{}'.format(args['action'])

        upload_data = {item["name"]: item["value"]
                       for item in args["fields"]}
        res = req.post(full_url, data=upload_data,
                       files={"file": image.read()})
        res.raise_for_status()

    def _post_to_subreddit(self, image_ids, post: RedditPost):
        body = {
            'sr': post.subreddit,
            'title': post.title,
            'sendreplies': True,  # TODO: customize?
        }

        self._apply_flair_to_body(body, post)

        if len(image_ids) > 1:
            return self._submit_gallery_post(body, image_ids)
        else:
            return self._submit_single_post(body, image_ids, post)

    def _apply_flair_to_body(self, body: dict, post: RedditPost):
        if post.flair:
            flair_id, flair_text = self._find_flair(post.subreddit, post.flair)
            if flair_id:
                body['flair_id'] = flair_id
                body['flair_text'] = flair_text
            else:
                raise Exception(f'Failed to find flair {post.flair}')

    def _find_flair(self, sr: str, text: str):
        if text:
            print(f'Searching for flair: {text}')
            res = req.get(f'{self.baseURL}/r/{sr}/api/link_flair',
                          headers=self.headers)
            res.raise_for_status()
            data = res.json()

            for flair in data:
                flair_text = flair['text']
                if text.lower() in flair_text.lower():
                    return flair['id'], flair_text

        return None, None

    def _submit_gallery_post(self, body: dict, image_ids):
        body['kind'] = 'self'
        body['items'] = [{'media_id': i, 'caption': '',
                          'outbound_url': ''} for i in image_ids]
        res = req.post(
            self.baseURL + '/api/submit_gallery_post.json?api_type=json', json=body, headers=self.headers)

        data = self._return_data_if_no_error(res)
        return data['id']  # , data['url']

    def _submit_single_post(self, body: dict, image_ids, post: RedditPost):
        if len(image_ids) > 0:
            body['kind'] = 'image'
            body['url'] = f'{self.asset_url}/rte_images/{image_ids[0]}',
        else:
            body['kind'] = 'self'
            body['text'] = post.text

        res = req.post(self.baseURL + '/api/submit?api_type=json',
                       body, headers=self.headers)
        data = self._return_data_if_no_error(res)

        if 'url' in data:
            return data['name']  # , data['url']

        return None  # , None


# ----------------- Testing -----------------
# load_dotenv()

# client = RedditClient(os.getenv('REDDIT_TOKEN'))

# test_sr = [{'sr': 'kumi_yada', 'flair': 'test'}]

# ###
# single_image_post = {
#     'title': 'Test submit single image post',
#     'text': 'Should be ignored',
#     'images': ['images/pixel-ina.png'],
# }
# client.submit_post(single_image_post, test_sr)
# ###
# multi_image_post = {
#     'title': 'Test submit multiple image post',
#     'text': 'Should be a comment\n\nNew line',
#     'images': ['images'],
# }
# client.submit_post(multi_image_post, test_sr)
# ###
# text_post = {
#     'title': 'Test text post',
#     'text': 'Should be the text of the post\n\n**New line**',
#     'images': [],
# }
# client.submit_post(text_post, test_sr)
# ###
